File: src/tts_sidecar/exit_codes.py
"""
Códigos de salida del CLI — contrato público congelado.

Un orquestador distingue las causas sin parsear texto.
La tabla completa vive aquí para que ningún módulo del paquete
necesite importar de cli.py y ningún sitio redefina una constante.

Criterio de asignación (clase de causa + siguiente llamada del consumidor):
  0  éxito
  1  error genérico (incluye chequeos fallidos de doctor)
  2  entrada inválida (texto vacío, nombre ilegal, colisión, uso incorrecto)
  3  recurso no encontrado voz o audio
  4  modelo no provisionado (ejecutar 'setup')
  5  daemon inalcanzable o no gestionable
  6  conflicto de estado (recurso ocupado, colisión de voz, puerto daemón en uso)
  7  operación no aplicable al contexto actual (sólo lectura, plataforma no soportada)
  8  precondición de entorno incumplida (credenciales, red, permisos, disco)
  130 interrupción por el usuario (128 + SIGINT)
"""

EXIT_OK = 0
EXIT_ERROR = 1
EXIT_INVALID_INPUT = 2
EXIT_NOT_FOUND = 3
EXIT_MODEL_MISSING = 4
EXIT_DAEMON_UNREACHABLE = 5
EXIT_STATE_CONFLICT = 6
EXIT_NOT_APPLICABLE = 7
EXIT_PRECONDITION_FAILED = 8
EXIT_INTERRUPTED = 130


class CliError(BaseException):
    """Fallo de la CLI. Hereda de BaseException (no Exception) para que
    ningún except Exception de dominio la capture y la degrade a 1.

    Atributos:
        code: código de salida del contrato público (EXIT_*)
        reason: razón breve y programática (p. ej. 'usage_error', 'credentials')
        message: mensaje humano legible (se imprime a stderr)
    """

    def __init__(self, code: int, reason: str, message: str) -> None:
        self.code = code
        self.reason = reason
        self.message = message
        super().__init__(message)

File: src/exit_codes.rs
use thiserror::Error;

#[derive(Error, Debug, Clone, Copy, PartialEq, Eq)]
pub enum ExitCode {
    #[error("Éxito")]
    Ok = 0,
    #[error("Error genérico")]
    Error = 1,
    #[error("Entrada inválida")]
    InvalidInput = 2,
    #[error("Recurso no encontrado")]
    NotFound = 3,
    #[error("Modelo no provisionado (ejecutar 'setup')")]
    ModelMissing = 4,
    #[error("Daemon inalcanzable o no gestionable")]
    DaemonUnreachable = 5,
    #[error("Conflicto de estado")]
    StateConflict = 6,
    #[error("Operación no aplicable al contexto actual")]
    NotApplicable = 7,
    #[error("Precondición de entorno incumplida")]
    PreconditionFailed = 8,
    #[error("Fallo del pipeline de traducción")]
    TranslationFailed = 9,
    #[error("Fallo del pipeline de transcripción")]
    TranscriptionFailed = 10,
    #[error("Interrupción por el usuario")]
    Interrupted = 130,
}

impl ExitCode {
    pub fn code(&self) -> i32 {
        *self as i32
    }
}

#[derive(Error, Debug)]
#[error("{message}")]
pub struct CliError {
    pub code: ExitCode,
    pub reason: String,
    pub message: String,
}

impl CliError {
    pub fn new(code: ExitCode, reason: impl Into<String>, message: impl Into<String>) -> Self {
        Self {
            code,
            reason: reason.into(),
            message: message.into(),
        }
    }

    pub fn invalid_input(reason: impl Into<String>, message: impl Into<String>) -> Self {
        Self::new(ExitCode::InvalidInput, reason, message)
    }

    pub fn not_found(reason: impl Into<String>, message: impl Into<String>) -> Self {
        Self::new(ExitCode::NotFound, reason, message)
    }
}
